Story/cite 95

MaulikLim opened the pull request:
# Guidelines for Pull Requests

If you haven't yet read our code review guidelines, please do so, You can find them [here](https://diging.atlassian.net/wiki/spaces/DIGING/pages/2256076801/Code+Review+Guidelines).

Please confirm the following by adding an x for each item (turn `[ ]` into `[x]`).

- [x] I have removed all code style changes that are not necessary (e.g. changing blanks across the whole file that don’t need to be changed, adding empty lines in parts other than your own code)
- [x] I am not making any changes to files that don’t have any effect (e.g. imports added that don’t need to be added)
- [x] I do not have any sysout statements in my code or commented out code that isn’t needed anymore
- [x] I am not reformatting any files in the wrong format or without cause. 
- [x] I am not changing file encoding or line endings to something else than UTF-8, LF
- [x] My pull request does not show an insane amount of files being changed although my ticket only requires a few files being changed
- [x] I have added Javadoc/documentation where appropriate
- [x] I have added test cases where appropriate
- [x] I have explained any part of my code/implementation decisions that is not be self-explanatory

## Please provide a brief description of your ticket 
(you can copy the ticket if it hasn't changed) 

[CITE-135] - Search author authority pop needs current group tab
[CITE-95] - In the authority popup for editing a citation, database queries should also search for last name only
  
## Anything else the reviewer needs to know?

This is a cumulative merge request for CITE-135 and CITE-95. Review comments from previous PR for CITE-135 have been addressed.
 [Previous PR](https://github.com/diging/citesphere/pull/162/files)


[CITE-135]: https://diging.atlassian.net/browse/CITE-135

diging-jenkins commented:
Can one of the admins verify this patch?

diging-jenkins commented:
Can one of the admins verify this patch?

jdamerow commented on citesphere/src/main/java/edu/asu/diging/citesphere/core/repository/custom/AuthorityRepository.java:
```diff
@@ -21,5 +22,9 @@
             @Param("firstName") String firstName, @Param("lastName") String lastName);
+    
+    @Query("select a from AuthorityEntry a where :group member of a.groups and a.name like %:firstName% and a.name like %:lastName% order by a.name")
+    public List<IAuthorityEntry> findByGroupAndFirstNameAndLastName(@Param("group") Long group,
```
have you tried using the automatic query generation here instead of writing the query?

MaulikLim replied:
When I tried to search for a translation for "member of", I didn't find any. Turns out "Containing" also works for this.

jdamerow commented on citesphere/src/main/java/edu/asu/diging/citesphere/core/service/impl/AuthorityService.java:
```diff
@@ -165,13 +164,17 @@ public IAuthorityEntry find(String id) {
+    @Override
+    public List<IAuthorityEntry> findByName(IUser user, String firstName, String lastName) {
         List<IAuthorityEntry> results = authorityRepository.findByUsernameAndNameContainingAndNameContainingOrderByName(
-                user.getUsername(), firstName, lastName, paging);
```
is there a reason you got rid of the pagination here?

jdamerow commented on citesphere/src/main/java/edu/asu/diging/citesphere/web/user/authorities/AuthorityEntryController.java:
```diff
@@ -51,15 +55,47 @@
+                    .filter(lastNameEntry -> !userEntryIds.contains(lastNameEntry.getId()))
+                    .collect(Collectors.toList());
+            userEntries.addAll(lastNameEntries);
+        }
```
this doesn't seem like a good approach. If there are 1000 results, then all 1000 results will be loaded into memory. The pagination needs to stay. I would just after you retrieved the page for first/last name search, check if there are fewer results than page size and if so fill in the rest with the ""/last name search. You can keep track of the pages for ""/last name search by looking at the past page (and how big the difference to total pages of first/last name search is).

jdamerow commented on citesphere/src/main/java/edu/asu/diging/citesphere/web/user/authorities/AuthorityEntryController.java:
```diff
@@ -51,15 +55,47 @@
+                    .filter(lastNameEntry -> !groupEntryIds.contains(lastNameEntry.getId()))
+                    .collect(Collectors.toList());
+            groupEntries.addAll(lastNameEntries);
+        }
```
see above

jdamerow commented on citesphere/src/main/java/edu/asu/diging/citesphere/web/user/authorities/AuthorityEntryController.java:
```diff
@@ -164,4 +200,12 @@
         return new ResponseEntity<IAuthorityEntry>(entry, HttpStatus.OK);
     }
 
+    private <T> List<T> getPage(List<T> entries, int page, int pageSize) {
```
is there a reason this is generic?

MaulikLim replied:
Doesn't have any other use over here. But it seemed that method could be a generic one.

MaulikLim commented:
Addressed review comments: reworked the pagination for search, used Spring JPA query instead of a custom one. Added Javadocs for AuthorityService methods as there are multiple similar methods.

jdamerow commented on citesphere/src/test/java/edu/asu/diging/citesphere/core/service/impl/AuthorityServiceTest.java:
```diff
@@ -94,23 +95,35 @@ private void initAuthorityEntries() {
+        Set<Long> groups;
+        
+        groups = new HashSet<>();
+        groups.add(123456L);
```
use the variable you defined above here

jdamerow replied:
same for the instances below

jdamerow commented:
Make it so, Jenkins.

diging-jenkins commented:
Congratulations, build succeeded!

MaulikLim commented on citesphere/src/main/java/edu/asu/diging/citesphere/core/service/IAuthorityService.java:
```diff
@@ -33,12 +35,61 @@ AuthoritySearchResult searchAuthorityEntries(IUser user, String firstName, Strin
 
-    List<IAuthorityEntry> findByName(IUser user, String firstName, String lastName, int page, int pageSize);
+    /**
+     * Finds the authorities which are created or imported by the given user whose
```
I'll correct it. The "whose" was for authorities. It sounded correct when I read it since I already knew and didn't give a second thought about it.

jdamerow commented:
- this seems to work fine if the first and last name actually exist (e.g. searching for "marie curie" brings up two results marie curie and pierre curie), but if the first name does not exist, it brings up no results (searching for "klaus curie" brings up 0 results although it should show the two previous ones).
- not sure if it this has been introduced in this ticket but, opening the search modal several times, append the group name again and again
- ![image](https://user-images.githubusercontent.com/8881141/128082013-18b065c1-c054-4ae7-86f2-f74439ce395f.png)

MaulikLim commented:
- Corrected Javadocs and code comments
- Fixed bug where authorities didn't show up in case the first name does not exist in database for any authority but the last name does
- Fixed the bug about duplicating group names in search pop-up for conceptpower when searched for multiple times.

diging-jenkins commented:
Congratulations, build succeeded!

jdamerow commented:
Resolve conflicts please

MaulikLim commented:
Resolved conflicts

diging-jenkins commented:
Congratulations, build succeeded!

jdamerow commented on citesphere/src/main/java/edu/asu/diging/citesphere/core/service/IAuthorityService.java:
```diff
@@ -41,12 +43,60 @@ AuthoritySearchResult searchAuthorityEntries(IUser user, String firstName, Strin
+     */
+    Page<IAuthorityEntry> findByGroupAndLastNameAndExcludingFirstName(Long groupId, String firstName, String lastName, int page, int pageSize);
+    
+    int getTotalGroupAuthoritiesPages(Long groupId, String firstName, String lastName, int pageSize);
```
get total methods need javadoc. It needs to explicitly. state that the total is only searching for first or last name. This is not intuitive or recognizable from the method name.

jdamerow commented:
Make it so, Jenkins.

diging-jenkins commented:
Congratulations, build succeeded!

jdamerow commented:
otherwise should  be good to go

diging-jenkins commented:
Congratulations, build succeeded!

jdamerow commented:
Make it so, Jenkins.

diging-jenkins commented:
Congratulations, build succeeded!
